collect_app/src/main/java/org/odk/collect/android/widgets/QuestionWidget.java

```diff
@@ -849,33 +809,38 @@ protected void deleteMediaFromContentProvider(Uri media) {
      * and if it's from Google Drive and not cached yet, we'll retrieve it using network.
      * This may take a long time.
      *
-     * @param selectedFile uri of the selected audio
+     * @param selectedFile uri of the selected media (audio/video/image/arbitrary file)
      */
     protected void saveChosenFile(Uri selectedFile) {
-        String extension = ContentResolverHelper.getFileExtensionFromUri(getContext(), selectedFile);
-        String instanceFolder = getFormController().getInstanceFile().getParent();
-        String destPath = instanceFolder + File.separator + System.currentTimeMillis() + extension;
-
-        try {
-            File chosenFile = MediaUtils.getFileFromUri(getContext(), selectedFile, MediaStore.Images.Media.DATA);
-            if (chosenFile != null) {
-                final File newFile = new File(destPath);
-                FileUtils.copyFile(chosenFile, newFile);
-                if (this instanceof BinaryWidget) {
-                    ((BinaryWidget) this).setBinaryData(newFile);
-                }
-                saveAnswersForCurrentScreen();
-            } else {
-                Timber.e("Could not receive chosen file");
-                ToastUtils.showShortToastInMiddle(R.string.error_occured);
-            }
-        } catch (GDriveConnectionException e) {
-            Timber.e("Could not receive chosen file due to connection problem");
-            ToastUtils.showLongToastInMiddle(R.string.gdrive_connection_exception);
-        }
-    }
 
-    private void refreshCurrentView() {
+        // show dialog
+        ((FormEntryActivity) getContext()).showDialog(SAVING_DIALOG);
+
+        ChosenFileSaver.createObservable(getContext(), selectedFile)
```

Comment:
Good stuff! Love to see RxJava at work to make async stuff easier to understand and process! I have some ideas about how this can be optimized so I am working on it to share!

Reply:
I am glad that you liked it @jd-alexander.

Reply:
Hey @shobhitagarwal1612 so I made some changes over on this branch here! Not quite done as yet but just wanted to give you a solid idea of the direction I think the `ChosenFileSaver` can go in! Check it out. Will make the relevant comments in the code here!
https://github.com/jd-alexander/collect/commit/88f7df1e04c512a4ec0227d5f09d502147b9fbcf